# ProteinCompletion

This repository contains the code for our ISMB 2026 submission, "Adjacent Token Prediction for Protein Sequence Motif Scaffolding." 

## Overview

Relevant executable scripts
- `train.py`: full training run from scratch with checkpointing
- `eval-completion.py`: motif scaffolding experiments
- `StructureEvaluation/structure_evaluator_driver.py`: main driver for running structure prediction and evaluation on generated sequences
- `StructureEvaluation/gen_combined_figs_driver.py`: driver for generating combined (multi-model) structure evaluation figures

Relevant non-executable scripts
- `utils/model_base.py`: defines all model components except prediction head
- `utils/model_bidirectional.py`: novel bidirectional model using ATP; increases size of LM head
- `utils/model_esmlike.py`: bidirectional model based on ESM3
- `utils/data.py`: all PyTorch `Dataset` definitions for preprocessing input data
- `utils/mask.py`: custom causal mask generation for ATP
- `utils/config.py`: defines `BaseConfig`, setting default values; config json files will always override these settings
- `utils/utils.py`: all other utility functions, notably including model and tokenizer loading
- `StructureEvaluation/structure_evaluator.py`: wraps the structure prediction model and computes pTM and pLDDT metrics
- `StructureEvaluation/generated_parser.py`: utilities for parsing generated and baseline sequence TSVs
- `StructureEvaluation/figure_generator.py`: plotting utilities for structure evaluation figures

## Usage

### Training
Training requires a single GPU with at least 48GB of VRAM. First, clone this repo. Within the cloned repo directory, download [UniProt SWISS-PROT](https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz), decompress, and move to a new subdirectory `/data`. Run the command below. We performed our training runs in increments of 24 hours with checkpointing. Our checkpoint loading system is slow but deterministic.
```
python train.py --max-samples 600000 --epochs 5 --save-every 20000 --save ./models/
```
Useful arguments:
- `--ckpt <filepath>`: Specify a checkpoint to load.
- `--model_type <bidirectional|esmlike>`: Specify whether you would like to train an ATP model or an ESM-like model.
- `--train <filepath>`: specify a different filename or location for your dataset.

### Experiments
To perform our motif scaffolding experiments, run the command below, specifying your own checkpoint (e.g., `train-step220000`) and model type. Results will be saved to `out.tsv`.
```
python eval-completion.py --weights ./weights/<checkpoint>.pt --model_type <bidirectional|esmlike>
```
Predicted secondary structures require a valid ESM3 API key to generate. We implement our secondary structure generation and output data generation in `StructureEvaluation/structure_evaluator_driver.py`. To run it yourself, edit `api_key` to store a valid API key and `tsv_path` to store the path to your output file from the previous step. You may then run the script without arguments. Our pLDDT and pTM figures shown in the paper may be generated by a similar procedure using `StructureEvaluation/gen_combined_figs_driver.py`

